In [1]:
import numpy as np
import pandas as pd

In [3]:
pd.set_option('display.max_columns',None)

In [4]:
df = pd.read_csv("/content/data.csv")
df.head(20)

id diagnosis  radius_mean  texture_mean  perimeter_mean  area_mean  \
0     842302         M        17.99         10.38          122.80     1001.0   
1     842517         M        20.57         17.77          132.90     1326.0   
2   84300903         M        19.69         21.25          130.00     1203.0   
3   84348301         M        11.42         20.38           77.58      386.1   
4   84358402         M        20.29         14.34          135.10     1297.0   
5     843786         M        12.45         15.70           82.57      477.1   
6     844359         M        18.25         19.98          119.60     1040.0   
7   84458202         M        13.71         20.83           90.20      577.9   
8     844981         M        13.00         21.82           87.50      519.8   
9   84501001         M        12.46         24.04           83.97      475.9   
10    845636         M        16.02         23.24          102.70      797.8   
11  84610002         M        15.78         17.89          103.60      781.0   
12    846226         M        19.17         24.80          132.40     1123.0   
13    846381         M        15.85         23.95          103.70      782.7   
14  84667401         M        13.73         22.61           93.60      578.3   
15  84799002         M        14.54         27.54           96.73      658.8   
16    848406         M        14.68         20.13           94.74      684.5   
17  84862001         M        16.13         20.68          108.10      798.8   
18    849014         M        19.81         22.15          130.00     1260.0   
19   8510426         B        13.54         14.36           87.46      566.3   

    smoothness_mean  compactness_mean  concavity_mean  concave points_mean  \
0           0.11840           0.27760         0.30010              0.14710   
1           0.08474           0.07864         0.08690              0.07017   
2           0.10960           0.15990         0.19740              0.12790   
3           0.14250           0.28390         0.24140              0.10520   
4           0.10030           0.13280         0.19800              0.10430   
5           0.12780           0.17000         0.15780              0.08089   
6           0.09463           0.10900         0.11270              0.07400   
7           0.11890           0.16450         0.09366              0.05985   
8           0.12730           0.19320         0.18590              0.09353   
9           0.11860           0.23960         0.22730              0.08543   
10          0.08206           0.06669         0.03299              0.03323   
11          0.09710           0.12920         0.09954              0.06606   
12          0.09740           0.24580         0.20650              0.11180   
13          0.08401           0.10020         0.09938              0.05364   
14          0.11310           0.22930         0.21280              0.08025   
15          0.11390           0.15950         0.16390              0.07364   
16          0.09867           0.07200         0.07395              0.05259   
17          0.11700           0.20220         0.17220              0.10280   
18          0.09831           0.10270         0.14790              0.09498   
19          0.09779           0.08129         0.06664              0.04781   

    symmetry_mean  fractal_dimension_mean  radius_se  texture_se  \
0          0.2419                 0.07871     1.0950      0.9053   
1          0.1812                 0.05667     0.5435      0.7339   
2          0.2069                 0.05999     0.7456      0.7869   
3          0.2597                 0.09744     0.4956      1.1560   
4          0.1809                 0.05883     0.7572      0.7813   
5          0.2087                 0.07613     0.3345      0.8902   
6          0.1794                 0.05742     0.4467      0.7732   
7          0.2196                 0.07451     0.5835      1.3770   
8          0.2350                 0.07389     0.3063      1.0020   
9         

In [31]:
df.drop('diagnosis', axis=1).skew().sort_values(ascending=False)

area_se                    5.447186
concavity_se               5.110463
fractal_dimension_se       3.923969
perimeter_se               3.443615
radius_se                  3.088612
smoothness_se              2.314450
symmetry_se                2.195133
compactness_se             1.902221
area_worst                 1.859373
fractal_dimension_worst    1.662579
texture_se                 1.646444
area_mean                  1.645732
compactness_worst          1.473555
concave points_se          1.444678
symmetry_worst             1.433928
concavity_mean             1.401180
fractal_dimension_mean     1.304489
compactness_mean           1.190123
concave points_mean        1.171180
concavity_worst            1.150237
perimeter_worst            1.128164
radius_worst               1.103115
perimeter_mean             0.990650
radius_mean                0.942380
symmetry_mean              0.725609
texture_mean               0.650450
texture_worst              0.498321
concave points_worst       0.492616
smoothness_mean            0.456324
smoothness_worst           0.415426
dtype: float64

In [5]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 569 entries, 0 to 568
Data columns (total 33 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   id                       569 non-null    int64  
 1   diagnosis                569 non-null    object 
 2   radius_mean              569 non-null    float64
 3   texture_mean             569 non-null    float64
 4   perimeter_mean           569 non-null    float64
 5   area_mean                569 non-null    float64
 6   smoothness_mean          569 non-null    float64
 7   compactness_mean         569 non-null    float64
 8   concavity_mean           569 non-null    float64
 9   concave points_mean      569 non-null    float64
 10  symmetry_mean            569 non-null    float64
 11  fractal_dimension_mean   569 non-null    float64
 12  radius_se                569 non-null    float64
 13  texture_se               569 non-null    float64
 14  perimeter_se             5

In [6]:
df.isnull().sum()

id                           0
diagnosis                    0
radius_mean                  0
texture_mean                 0
perimeter_mean               0
area_mean                    0
smoothness_mean              0
compactness_mean             0
concavity_mean               0
concave points_mean          0
symmetry_mean                0
fractal_dimension_mean       0
radius_se                    0
texture_se                   0
perimeter_se                 0
area_se                      0
smoothness_se                0
compactness_se               0
concavity_se                 0
concave points_se            0
symmetry_se                  0
fractal_dimension_se         0
radius_worst                 0
texture_worst                0
perimeter_worst              0
area_worst                   0
smoothness_worst             0
compactness_worst            0
concavity_worst              0
concave points_worst         0
symmetry_worst               0
fractal_dimension_worst      0
Unnamed: 32                569
dtype: int64

In [7]:
df.shape

(569, 33)

In [9]:
df.drop(columns =['id', 'Unnamed: 32'],inplace= True)

In [12]:
df['diagnosis'].nunique()

2

In [14]:
df['diagnosis'].unique()

array(['M', 'B'], dtype=object)

In [15]:
df['diagnosis'] = df['diagnosis'].map({'M':1,'B':0})

In [17]:
df.value_counts

<bound method DataFrame.value_counts of      diagnosis  radius_mean  texture_mean  perimeter_mean  area_mean  \
0            1        17.99         10.38          122.80     1001.0   
1            1        20.57         17.77          132.90     1326.0   
2            1        19.69         21.25          130.00     1203.0   
3            1        11.42         20.38           77.58      386.1   
4            1        20.29         14.34          135.10     1297.0   
..         ...          ...           ...             ...        ...   
564          1        21.56         22.39          142.00     1479.0   
565          1        20.13         28.25          131.20     1261.0   
566          1        16.60         28.08          108.30      858.1   
567          1        20.60         29.33          140.10     1265.0   
568          0         7.76         24.54           47.92      181.0   

     smoothness_mean  compactness_mean  concavity_mean  concave points_mean  \
0            0.11840           0.27760         0.30010              0.14710   
1            0.08474           0.07864         0.08690              0.07017   
2            0.10960           0.15990         0.19740              0.12790   
3            0.14250           0.28390         0.24140              0.10520   
4            0.10030           0.13280         0.19800              0.10430   
..               ...               ...             ...                  ...   
564          0.11100           0.11590         0.24390              0.13890   
565          0.09780           0.10340         0.14400              0.09791   
566          0.08455           0.10230         0.09251              0.05302   
567          0.11780           0.27700         0.35140              0.15200   
568          0.05263           0.04362         0.00000              0.00000   

     symmetry_mean  fractal_dimension_mean  radius_se  texture_se  \
0           0.2419                 0.07871     1.0950      0.9053   
1           0.1812                 0.05667     0.5435      0.7339   
2           0.2069                 0.05999     0.7456      0.7869   
3           0.2597                 0.09744     0.4956      1.1560   
4           0.1809                 0.05883     0.7572      0.7813   
..             ...                     ...        ...         ...   
564         0.1726                 0.05623     1.1760      1.2560   
565         0.1752                 0.05533     0.7655      2.4630   
566         0.1590                 0.05648     0.4564      1.0750   
567         0.2397                 0.07016     0.7260      1.5950   
568         0.1587                 0.05884     0.3857      1.4280   

     perimeter_se  area_se  smoothness_se  compactness_se  concavity_se  \
0           8.589   153.40       0.006399         0.04904       0.05373   
1           3.398    74.08       0.005225         0.01308       0.01860   
2           4.585    94.03       0.006150         0.04006       0.03832   
3           3.445    27.23       0.009110         0.07458       0.05661   
4           5.438    94.44       0.011490         0.02461       0.05688   
..            ...      ...            ...             ...           ...   
564         7.673   158.70       0.010300         0.02891       0.05198   
565         5.203    99.04       0.005769         0.02423       0.03950   
566         3.425    48.55       0.005903         0.03731       0.04730   
567         5.772    86.22       0.006522         0.06158       0.07117   
568         2.548    19.15       0.007189         0.00466       0.00000   

     concave points_se  symmetry_se  fractal_dimension_se  radius_worst  \
0              0.01587      0.03003              0.006193        25.380   
1              0.01340      0.01389              0.003532        24.990   
2              0.02058      0.02250              0.004571        23.570   
3              0.01867      0.05963              0.009208        14.910   
4              0.01885      0.01756      

In [19]:
df['diagnosis'].value_counts()

diagnosis
0    357
1    212
Name: count, dtype: int64

In [20]:
X = df.drop(columns = 'diagnosis')
y = df['diagnosis']

In [21]:
from sklearn.model_selection import train_test_split
X_train,X_test,y_train,y_test = train_test_split(X,y,test_size=0.2,random_state=42)

In [22]:
from sklearn.feature_selection import RFE
from sklearn.tree import DecisionTreeClassifier

dt = DecisionTreeClassifier(
    random_state = 42,
    criterion = 'gini',
    max_depth = 5,
    min_samples_split = 10,
    min_samples_leaf = 5,
    max_features = 'sqrt'
)

rfe = RFE(
    estimator = dt,
    n_features_to_select = 10
)

rfe.fit(X_train,y_train)

selected_features = X_train.columns[rfe.support_]
ranking = pd.Series(rfe.ranking_, index=X_train.columns).sort_values()

print("Selected Features:\n", selected_features)
print("\nFeature Rankings:\n", ranking)

Selected Features:
 Index(['fractal_dimension_se', 'radius_worst', 'texture_worst',
       'perimeter_worst', 'area_worst', 'smoothness_worst', 'concavity_worst',
       'concave points_worst', 'symmetry_worst', 'fractal_dimension_worst'],
      dtype='object')

Feature Rankings:
 smoothness_worst            1
concavity_worst             1
concave points_worst        1
area_worst                  1
radius_worst                1
texture_worst               1
symmetry_worst              1
fractal_dimension_worst     1
perimeter_worst             1
fractal_dimension_se        1
compactness_worst           2
symmetry_se                 3
compactness_se              4
concave points_se           5
area_se                     6
smoothness_se               7
concavity_se                8
texture_se                  9
radius_se                  10
concave points_mean        11
perimeter_se               12
fractal_dimension_mean     13
symmetry_mean              14
compactness_mean           1

In [23]:
y_train_pred = rfe.predict(X_train)
y_val_pred = rfe.predict(X_test)

In [36]:
print("Selected Features:\n", selected_features)
print("\nFeature Rankings:\n", ranking)

Selected Features:
 Index(['fractal_dimension_se', 'radius_worst', 'texture_worst',
       'perimeter_worst', 'area_worst', 'smoothness_worst', 'concavity_worst',
       'concave points_worst', 'symmetry_worst', 'fractal_dimension_worst'],
      dtype='object')

Feature Rankings:
 smoothness_worst            1
concavity_worst             1
concave points_worst        1
area_worst                  1
radius_worst                1
texture_worst               1
symmetry_worst              1
fractal_dimension_worst     1
perimeter_worst             1
fractal_dimension_se        1
compactness_worst           2
symmetry_se                 3
compactness_se              4
concave points_se           5
area_se                     6
smoothness_se               7
concavity_se                8
texture_se                  9
radius_se                  10
concave points_mean        11
perimeter_se               12
fractal_dimension_mean     13
symmetry_mean              14
compactness_mean           1

In [24]:
from sklearn.metrics import accuracy_score
train_accuracy = accuracy_score(y_train, y_train_pred)
val_accuracy = accuracy_score(y_test, y_val_pred)

In [26]:
print(f"\nTraining Accuracy: {train_accuracy:.4f}")
print(f"Validation Accuracy: {val_accuracy:.4f}")


Training Accuracy: 0.9626
Validation Accuracy: 0.9649


In [33]:
from sklearn.metrics import (
    accuracy_score,
    precision_score,
    recall_score,
    f1_score
)

In [27]:
from sklearn.tree import DecisionTreeClassifier
from sklearn.metrics import accuracy_score

# Decision Tree (same params as RFE)
dt_full = DecisionTreeClassifier(
    random_state=42,
    criterion='gini',
    max_depth=5,
    min_samples_split=10,
    min_samples_leaf=5,
    max_features='sqrt'
)

dt_full.fit(X_train, y_train)

# Predictions
y_train_pred_full = dt_full.predict(X_train)
y_val_pred_full = dt_full.predict(X_test)

# Accuracy
train_acc_full = accuracy_score(y_train, y_train_pred_full)
val_acc_full = accuracy_score(y_test, y_val_pred_full)

In [28]:
# Predictions using RFE model
y_train_pred_rfe = rfe.predict(X_train)
y_val_pred_rfe = rfe.predict(X_test)

# Accuracy
train_acc_rfe = accuracy_score(y_train, y_train_pred_rfe)
val_acc_rfe = accuracy_score(y_test, y_val_pred_rfe)

In [29]:
comparison = pd.DataFrame({
    'Model': ['Decision Tree (All Features)', 'Decision Tree (RFE Features)'],
    'Training Accuracy': [train_acc_full, train_acc_rfe],
    'Validation Accuracy': [val_acc_full, val_acc_rfe]
})

comparison

Model  Training Accuracy  Validation Accuracy
0  Decision Tree (All Features)           0.951648             0.956140
1  Decision Tree (RFE Features)           0.962637             0.964912

“Using Recursive Feature Elimination with a constrained Decision Tree slightly improved validation accuracy (95.6% → 96.5%) while reducing the feature set. This indicates substantial feature redundancy in the dataset. However, the improvement is marginal, and all features were retained for Random Forest training to leverage feature bagging.

# Breast Cancer Diagnosis – Decision Tree & Random Forest (EDA + Modeling)

## Project Overview
This project focuses on performing **Exploratory Data Analysis (EDA)** and building **tree-based classification models** to predict whether a breast tumor is **malignant or benign** using the Breast Cancer Wisconsin (Diagnostic) dataset.

The primary goal is to understand how **Decision Trees and Random Forests** behave with respect to feature distributions, correlation, feature selection, and overfitting, rather than applying unnecessary preprocessing.

---

## Dataset Information

- **Dataset Name:** Breast Cancer Wisconsin (Diagnostic)
- **Source:** UCI Machine Learning Repository  
- **Dataset Link:**  
  https://archive.ics.uci.edu/ml/datasets/Breast+Cancer+Wisconsin+(Diagnostic)

- **Shape:** `(569, 31)`
- **Target Column:** `diagnosis`
  - `1` → Malignant  
  - `0` → Benign  

Each feature represents computed characteristics of cell nuclei obtained from digitized images of fine needle aspirate (FNA) biopsies.

---

## Data Cleaning & Preprocessing

The following steps were performed:

- Dropped the `id` column (non-informative identifier)
- Removed the column with only null values (`Unnamed: 32`)
- Encoded the target variable:
  - `M → 1`
  - `B → 0`
- No feature scaling or skewness correction was applied, as tree-based models are invariant to monotonic transformations

---

## Exploratory Data Analysis (EDA)

### 1. Target Class Distribution
- Benign (0): **357 samples**
- Malignant (1): **212 samples**

This represents a **mild class imbalance (~63% vs ~37%)**, which does not require resampling for tree-based models.

---

### 2. Feature Distribution by Class
- Boxplots were used to compare key features across benign and malignant classes
- Features such as `radius_mean`, `perimeter_mean`, `area_mean`, and `concavity_mean` showed clear separation
- This explains why tree-based models perform well on this dataset

---

### 3. Correlation Analysis
- Strong correlations were observed among groups of features (e.g., radius, perimeter, area)
- Correlated features were **not removed**, as:
  - Decision Trees select splits based on thresholds
  - Random Forests use feature bagging to reduce correlation impact

---

### 4. Skewness & Multicollinearity
- Skewness and multicollinearity were analyzed but **not corrected**
- These issues primarily affect linear models, not tree-based models
- Retaining correlated features helps Random Forests leverage feature diversity

---

## Feature Selection (RFE)

- **Recursive Feature Elimination (RFE)** was applied using a constrained Decision Tree
- Purpose: **feature relevance analysis**, not aggressive feature removal
- A subset of 10 important features was selected
- RFE was used only for analysis; all features were retained for Random Forest training

---

## Model Training & Evaluation

### Decision Tree
- Controlled using:
  - `max_depth`
  - `min_samples_split`
  - `min_samples_leaf`
  - `max_features`
- Training vs validation accuracy was analyzed across different tree depths to visualize overfitting

---

### RFE vs Full Feature Comparison (Decision Tree)

| Model | Training Accuracy | Validation Accuracy |
|-----|------------------|--------------------|
| Decision Tree (All Features) | 0.9516 | 0.9561 |
| Decision Tree (RFE Features) | 0.9626 | 0.9649 |

**Observation:**  
RFE slightly improved validation accuracy, indicating feature redundancy. However, the improvement is marginal, and all features were retained for ensemble models.

---

### Random Forest
- Used to reduce variance and improve generalization
- Performance evaluated using:
  - Accuracy
  - Precision
  - Recall
  - ROC-AUC score
- ROC curves were plotted to compare Decision Tree vs Random Forest performance

---

## Key Takeaways

- Tree-based models handle skewness and multicollinearity naturally
- Mild class imbalance does not require resampling
- Recursive Feature Elimination is useful for **analysis**, not mandatory feature removal
- Random Forests generalize better than single Decision Trees
- ROC-AUC and recall are more informative than accuracy alone for medical classification tasks

---

## Tools & Libraries
- Python
- pandas, numpy
- scikit-learn

---

## Conclusion
This project demonstrates a **methodical, model-aware EDA workflow** and highlights why Decision Trees and Random Forests are effective for structured medical datasets with correlated features.
